### ISSUE TITLE
CJoinElement.getColumnSelect():  SQL Server + CGridView + with() + named-scope + pagination causes SQL error 

### ISSUE THREAD

**Issue by qiangxue** [yiisoft/yii#225]
<b>What steps will reproduce the problem?</b>
1. Define two models e.g. &quot;Parent&quot; and &quot;Child&quot;.  Parent model has relation 'child'
2. define Parent named scopes() to reference Child using 'with', e.g.

```php
  public function scopes() {
    return array ( 
      'MyNamedScope'=>array(
          'condition' => '.....',
          'with'=> array(
              'child'=>array(
                'select'=>'<column list>',
                'condition'=>...,
              ),
           ),
       ),
    );
  }
```

3. create standard crud 'admin' view using CGridView of Parent.  $model is Parent.  set 

```php
  'dataProvider' => $model->MyNamedScope()->search(),
```

4.open the first page of Admin view.  everything is fine.
5. go to next page of view. 

<b>What is the expected output? What do you see instead?</b>
Expected is the 2nd page shows.
Actual is SQL Error that columns have been defined twice.  In my case Parent table has composite PK of 3 columns.  Each of the three columns is in the SELECT clause twice, on the 2nd page.  In the first page there is no such error.

```
[SQLSTATE[42000]: [Microsoft][SQL Server Native Client 10.0][SQL Server]The column 't0_c1' was specified multiple times for '__inner__'..]
```

<b>What version of the product are you using? On what operating system?</b>
Yii 1.1.7 with patch r3173 applied to CActiveFinder.php (CJoinElement.getColumnSelect() )

The problem appears to be in CJoinElement.getColumnSelect() in the section commented &quot;//add primary key selection if they are not selected&quot;.   

==> instead of checking for $selected[$name] of each PK column, it should check $selected [$ALIAS] since that is what $selected contains, column aliases not column names.

MS SQL Server 2008
Windows XP SP3
PHP 5.2.17 / Apache 5.2.17

<b>Please provide any additional information below.</b>

<p>Migrated from http://code.google.com/p/yii/issues/detail?id=2736</p><hr/><h2>earlier comments</h2>
<p><strong>jeremy.j.dunn said, at 2011-08-12T21:44:00.000Z:</strong></p>
Here is a patch that fixes the problem, in CActiveFinder.php
CJoinElement.getColumnSelect():

```php
// add primary key selection if they are not selected
if(is_string($this->_pkAlias) && !isset($selected[$this->_pkAlias]))
    $columns[]=$prefix.$schema->quoteColumnName($this->_table->primaryKey).' AS '.$schema->quoteColumnName($this->_pkAlias);
else if(is_array($this->_pkAlias))
{
  foreach($this->_table->primaryKey as $name) 
  //if(!isset($selected[$name]))    // -JJD
  if(!isset($selected[$this->_pkAlias[$name]]))   // + JJD
	$columns[]=$prefix.$schema->quoteColumnName($name).' AS '.$schema->quoteColumnName($this->_pkAlias[$name]);
}
```
<p><strong>qiang.xue said, at 2012-01-01T03:37:10.000Z:</strong></p>
<p>set for 1.1.10 milestone</p><p><strong>qiang.xue said, at 2012-01-01T03:37:35.000Z:</strong></p>
<p>set for 1.1.10 milestone</p>

**Comment by djagya** [comment]
@qiangxue i'm struggling with this issue. Could we maybe apply this patch, that was added jeremy.j.dunn?

```
CJoinElement::getColumnSelect()

// add primary key selection if they are not selected
if (is_string($this->_pkAlias) && !isset($selected[$this->_pkAlias])) {
    $columns[] = $prefix . $schema->quoteColumnName($this->_table->primaryKey) . ' AS ' . $schema->quoteColumnName($this->_pkAlias);
} elseif (is_array($this->_pkAlias)) {
    foreach ($this->_table->primaryKey as $name) {
        ---if (!isset($selected[$name])) {
        +++if(!isset($selected[$this->_pkAlias[$name]])) {
            $columns[] = $prefix . $schema->quoteColumnName($name) . ' AS ' . $schema->quoteColumnName($this->_pkAlias[$name]);
        }
    }
}
```

It's an easy and fast patch, that will help a lot. I can do that.

**Comment by samdark** [comment]
It would be great if you'll prepare a pull request for it containing both the patch and some unit tests checking it.

**Comment by djagya** [comment]
@samdark sorry, turned out, that this issue was already fixed in 1.1.16. This issue should be closed

**Comment by samdark** [comment]
Great. Thanks for your time.
